### ISSUE TITLE
Zero-Dates Cause Pulses To Fail

### ISSUE THREAD

**Issue by kurrestahlberg** [metabase/metabase#11665]
**Describe the bug**
Some of our pulses started failing with upgrade to 0.34.0 from 0.32.10 and we found that the system tries to format a date that has a null value. This is odd as the table does not have null values, only zero-date values (`0000-00-00`) with proper dates. As the zero-date values show as empty in question results we can only assume that the driver or some other part of the data retrieval process converts zero-dates to make sure they show as empty fields. The question itself works and produces expected results, it only fails when used in a pulse.

The preview of any pulses that have this will keep spinning forever, but what is more serious is that the background pulse sending job seems to crash and metabase needs to be restarted to resume pulse sending.

The problem can be worked around by changing `SELECT date_entry FROM table;` to `SELECT IFNULL(date_entry, "0000-00-00") as date_entry FROM table;` but this requires changing all questions to native SQL instead of using the question builder.

**Logs**
```
[94ada379-043b-4e91-93a3-cc2e48762250] 2020-01-09T10:54:39+01:00 DEBUG metabase.middleware.log GET /api/pulse/11 200 17.4 ms (7 DB calls) Jetty threads: 4/50 (17 idle, 0 queued) (81 total active threads) Queries in flight: 0
[94ada379-043b-4e91-93a3-cc2e48762250] 2020-01-09T10:54:39+01:00 DEBUG metabase.middleware.log GET /api/pulse/form_input 200 294.0 ms (0 DB calls) Jetty threads: 4/50 (17 idle, 0 queued) (81 total active threads) Queries in flight: 1
[94ada379-043b-4e91-93a3-cc2e48762250] 2020-01-09T10:54:44+01:00 ERROR metabase.middleware.log GET /api/pulse/preview_card_info/173 500 5.2 s (13 DB calls) 
{:message "temporal",
 :type java.lang.NullPointerException,
 :stacktrace
 ("java.base/java.util.Objects.requireNonNull(Unknown Source)"
  "java.base/java.time.format.DateTimeFormatter.formatTo(Unknown Source)"
  "java.base/java.time.format.DateTimeFormatter.format(Unknown Source)"
  "java_time.format$format.invokeStatic(format.clj:50)"
  "java_time.format$format.invoke(format.clj:43)"
  "--> pulse.render.datetime$reformat_temporal_str.invokeStatic(datetime.clj:14)"
  "pulse.render.datetime$reformat_temporal_str.invoke(datetime.clj:13)"
  "pulse.render.datetime$format_temporal_str.invokeStatic(datetime.clj:33)"
  "pulse.render.datetime$format_temporal_str.invoke(datetime.clj:16)"
  "pulse.render.body$fn__32751$format_cell__32756$fn__32757.invoke(body.clj:46)"
  "pulse.render.body$fn__32751$format_cell__32756.invoke(body.clj:43)"
  "pulse.render.body$fn__32824$query_results__GT_row_seq__32829$fn__32830$iter__32831__32835$fn__32836$iter__32869__32873$fn__32874.invoke(body.clj:97)"
  "pulse.render.table$render_table_body$iter__32665__32669$fn__32670$iter__32703__32707$fn__32708.invoke(table.clj:75)"
  "api.pulse$fn__65440.invokeStatic(pulse.clj:166)"
  "api.pulse$fn__65440.invoke(pulse.clj:159)"
  "middleware.auth$enforce_authentication$fn__48819.invoke(auth.clj:14)"
  "routes$fn__67566$fn__67567.doInvoke(routes.clj:56)"
  "middleware.exceptions$catch_uncaught_exceptions$fn__48775.invoke(exceptions.clj:104)"
  "middleware.exceptions$catch_api_exceptions$fn__48772.invoke(exceptions.clj:92)"
  "middleware.log$log_api_call$fn__48685$fn__48686.invoke(log.clj:178)"
  "middleware.log$log_api_call$fn__48685.invoke(log.clj:172)"
  "middleware.security$add_security_headers$fn__48736.invoke(security.clj:121)"
  "middleware.json$wrap_json_body$fn__48904.invoke(json.clj:64)"
  "middleware.json$wrap_streamed_json_response$fn__48922.invoke(json.clj:100)"
  "middleware.session$bind_current_user$fn__47354$fn__47355.invoke(session.clj:209)"
  "middleware.session$do_with_current_user.invokeStatic(session.clj:184)"
  "middleware.session$do_with_current_user.invoke(session.clj:177)"
  "middleware.session$bind_current_user$fn__47354.invoke(session.clj:208)"
  "middleware.session$wrap_current_user_id$fn__47339.invoke(session.clj:161)"
  "middleware.session$wrap_session_id$fn__47283.invoke(session.clj:125)"
  "middleware.auth$wrap_api_key$fn__48827.invoke(auth.clj:27)"
  "middleware.misc$maybe_set_site_url$fn__48800.invoke(misc.clj:56)"
  "middleware.misc$bind_user_locale$fn__48803.invoke(misc.clj:72)"
  "middleware.misc$add_content_type$fn__48788.invoke(misc.clj:28)"
  "middleware.misc$disable_streaming_buffering$fn__48811.invoke(misc.clj:87)")}
```

**To Reproduce**
1. Create a table with a date field like this: `date_entry date NOT NULL DEFAULT '0000-00-00'`
2. Add some entries, both with zero dates and actual dates
3. Create a question listing the data from the table
4. Create a pulse that sends the results in an email

**Expected behavior**
Results are sent via email, pulse job doesn't crash

**Information about your Metabase Installation:**

```json
{
  "browser-info": {
    "language": "fi",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:73.0) Gecko/20100101 Firefox/73.0",
    "vendor": ""
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.3+7",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.3",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.3+7",
    "os.name": "Linux",
    "os.version": "4.14.152-98.182.amzn1.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "mysql",
      "googleanalytics",
      "bigquery",
      "h2"
    ],
    "hosting-env": "elastic-beanstalk",
    "application-database": "postgres",
    "run-mode": "prod",
    "version": {
      "date": "2019-12-19",
      "tag": "v0.34.0",
      "branch": "release-0.34.x",
      "hash": "1335215"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
Mostly annoying but prevents us from adding more pulses unless a developer is available to convert the question to SQL and add the workaround

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by dacort** [comment]
Possibly an issue with MySQL null time values as well per https://discourse.metabase.com/t/timediff-or-null-time-field/8553

**Comment by Nico-P-FR** [comment]
Hello, we have same kind of problem (a MySQL query works, same query in a pulse fail).

These pulses were OK before our update to version 4.

`[d70bc86e-97ed-41bc-b558-699408cee3c6] 2020-02-03T11:46:51+01:00 ERROR metabase.middleware.log GET /api/pulse/preview_card_info/585 500 2.1 s (12 DB calls)
{:message "**Unrecognized token 'Text': was expecting ('true', 'false' or 'null')\n at [Source: (StringReader); line: 1, column: 5]"**,
:type com.fasterxml.jackson.core.JsonParseException,
:stacktrace
("com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1804)"
"com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:679)"
"com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:2835)"
"com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue(ReaderBasedJsonParser.java:1889)"
"com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:747)"
"cheshire.parse$parse.invokeStatic(parse.clj:88)"
"cheshire.parse$parse.invoke(parse.clj:86)"
"cheshire.core$parse_string.invokeStatic(core.clj:206)"
"cheshire.core$parse_string.invoke(core.clj:192)"
"cheshire.core$parse_string.invokeStatic(core.clj:202)"
"cheshire.core$parse_string.invoke(core.clj:192)"
"clojure.core$comp$fn__5807.invoke(core.clj:2569)"
"clojure.core$comp$fn__5807.invoke(core.clj:2569)"
"medley.core$map_keys$fn__1762$fn__1763.invoke(core.cljc:89)"
"clojure.lang.PersistentArrayMap.kvreduce(PersistentArrayMap.java:377)"
"clojure.core$fn__8437.invokeStatic(core.clj:6845)"
"clojure.core$fn__8437.invoke(core.clj:6830)"
"clojure.core.protocols$fn__8167$G__8162__8176.invoke(protocols.clj:175)"
"clojure.core$reduce_kv.invokeStatic(core.clj:6856)"
"clojure.core$reduce_kv.invoke(core.clj:6847)"
"medley.core$reduce_map.invokeStatic(core.cljc:70)"
"medley.core$reduce_map.invoke(core.cljc:67)"
"medley.core$map_keys.invokeStatic(core.cljc:89)"
"medley.core$map_keys.invoke(core.cljc:86)"
"--> pulse.render.body$column_name.invokeStatic(body.clj:71)"
"pulse.render.body$column_name.invoke(body.clj:67)"
"pulse.render.body$query_results__GT_header_row$iter__32026__32030$fn__32031.invoke(body.clj:86)"
"pulse.render.table$render_table_head$iter__31874__31878$fn__31879.invoke(table.clj:59)"
"api.pulse$fn__63391.invokeStatic(pulse.clj:166)"
"api.pulse$fn__63391.invoke(pulse.clj:159)"
"middleware.auth$enforce_authentication$fn__66128.invoke(auth.clj:14)"
"routes$fn__67481$fn__67482.doInvoke(routes.clj:56)"
"middleware.exceptions$catch_uncaught_exceptions$fn__66227.invoke(exceptions.clj:104)"
"middleware.exceptions$catch_api_exceptions$fn__66224.invoke(exceptions.clj:92)"
"middleware.log$log_api_call$fn__67857$fn__67858.invoke(log.clj:178)"
"middleware.log$log_api_call$fn__67857.invoke(log.clj:172)"
"middleware.security$add_security_headers$fn__66190.invoke(security.clj:121)"
"middleware.json$wrap_json_body$fn__67562.invoke(json.clj:64)"
"middleware.json$wrap_streamed_json_response$fn__67580.invoke(json.clj:100)"
"middleware.session$bind_current_user$fn__62721$fn__62722.invoke(session.clj:209)"
"middleware.session$do_with_current_user.invokeStatic(session.clj:184)"
"middleware.session$do_with_current_user.invoke(session.clj:177)"
"middleware.session$bind_current_user$fn__62721.invoke(session.clj:208)"
"middleware.session$wrap_current_user_id$fn__62706.invoke(session.clj:161)"
"middleware.session$wrap_session_id$fn__62650.invoke(session.clj:125)"
"middleware.auth$wrap_api_key$fn__66136.invoke(auth.clj:27)"
"middleware.misc$maybe_set_site_url$fn__67884.invoke(misc.clj:56)"
"middleware.misc$bind_user_locale$fn__67887.invoke(misc.clj:72)"
"middleware.misc$add_content_type$fn__67872.invoke(misc.clj:28)"
"middleware.misc$disable_streaming_buffering$fn__67895.invoke(misc.clj:87)")}`

**Comment by kurrestahlberg** [comment]
> Hello, we have same kind of problem (a MySQL query works, same query in a pulse fail).
> 
> These pulses were OK before our update to version 4. And the workaround suggest by @kurrestahlberg doesn't work for me (or I didn't succeed to use it)
> 
> `[d70bc86e-97ed-41bc-b558-699408cee3c6] 2020-02-03T11:46:51+01:00 ERROR metabase.middleware.log GET /api/pulse/preview_card_info/585 500 2.1 s (12 DB calls) {:message "**Unrecognized token 'Text': was expecting ('true', 'false' or 'null')\n at [Source: (StringReader); line: 1, column: 5]"**, :type com.fasterxml.jackson.core.JsonParseException, :stacktrace `

This looks to be similar (and possibly related) but different issue. The parser is failing to parse a boolean JSON value instead of crashing on date formatting. This would explain why the work-around doesn't help.

**Comment by mkopinsky** [comment]
I was able to work around this with `IFNULL(dateColumn, '')` - OP's suggested `IFNULL(dateColumn,'00-00-0000')` displayed the zeroes in the report.

**Comment by pan-christensen** [comment]
Another workaround is to remove the offending column. This does not require converting the question to SQL. However, it's not a good solution if you need the data in that column.

**Comment by rodrigotassinari** [comment]
I'm having the same issue (description and logs on #12278 -- closed as duplicate).

I could only manage for the question to work on Pulses and on Metabot by replacing the null dates with `'1970-01-01'` -- Redshift would not accept empty strings as a replacement. But I can't have 1970-01-01 appearing on the reports, so the proposed workaround is not sufficient for my case. Does anyone know another Redshift-compatible workaround?

**Comment by kemathur** [comment]
Hitting the same issue with NULL date fields.

**Comment by TrineFN** [comment]
We are experiencing the same issue with a Redshift DB. 

As a work-around I tried casted the timestamp to a string `to_char(timestamp, 'Month d, YYYY')`. After this I no longer get the error message `metabase.task.send-pulses Error sending Pulse X` and if I check the task history there is no failure for the send-pulse task. 
The spinning issue that @kurrestahlberg mentioned ("the preview of any pulses that have this will keep spinning forever...") is also gone when casting the timestamp to a string.
BUT emails are still not being sent. All I get now is `WARN metabase.email Failed to send email`. I cannot find any further error messages.

**Comment by TrineFN** [comment]
Adding to my previous comment:
With the work-around the pulse is triggered correctly to Slack, but emails are not send. 

I now created a new simple SQL question as a test (without timestamp) and setup an alert:
`select 
    1 as bjad`

This simple alert also fails with `WARN metabase.email Failed to send email`. 

We are having these pulse issues since we updated to v0.35.1 from v0.33.4

**Comment by flamber** [comment]
@TrineFN Perhaps you're seeing another issue, besides this? But I cannot reproduce your `select 1 as bjad` with an alert (not pulse), sent to email, I'm getting that email without errors.
There might be more details hiding at different levels, so try using `TRACE` or `DEBUG` logging:
https://www.metabase.com/docs/latest/operations-guide/log-configuration.html

**Comment by TrineFN** [comment]
@flambert Thanks! We did indeed see another issue, but one caused by us (while switching to 2-step-authentication...). We have fixed that now and with the workaround I mentioned for the null dates pulses and alerts are now triggered correctly. 

@rodrigotassinari The above mentioned workaround with casting the timestamp to a string might work for you too?

**Comment by rodrigotassinari** [comment]
I've just tested the above mentioned workaround (casting the timestamp to a string via `to_char`) and it also works for me. thanks @TrineFN !

**Comment by robdaemon** [comment]
Fix merged, targeted for 0.35.4 release.
